File: src/main/scala/mmlTests/a201701b/java/gpt4/TraceFactoryImpl.java
package mmlTests.a201701b.java.gpt4;

import mmlTests.a201701b.java.Event;
import mmlTests.a201701b.java.Trace;
import mmlTests.a201701b.java.TraceFactory;

import java.util.function.Supplier;
import java.util.stream.IntStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Optional;
import java.util.function.Supplier;
public class TraceFactoryImpl implements TraceFactory {

	@Override
	public <X> Trace<X> fromSuppliers(Supplier<Integer> sdeltaTime, Supplier<X> svalue, int size) {
		List<Event<X>> events = new ArrayList<>();
		int currentTime = 0;
		for (int i = 0; i < size; i++) {
			int deltaTime = sdeltaTime.get();
			X value = svalue.get();
			Event<X> event = new EventImpl<>(currentTime, value);
			events.add(event);
			currentTime += deltaTime;
		}
		return new TraceImpl<>(events.iterator());
	}

	@Override
	public <X> Trace<X> constant(Supplier<Integer> sdeltaTime, X value, int size) {
		Supplier<X> svalue = () -> value;
		return fromSuppliers(sdeltaTime, svalue, size);
	}

	@Override
	public <X> Trace<X> discrete(Supplier<X> svalue, int size) {
		Supplier<Integer> sdeltaTime = new Supplier<Integer>() {
			private int currentTime = 0;

			@Override
			public Integer get() {
				int deltaTime = currentTime;
				currentTime++;
				return deltaTime;
			}
		};
		return fromSuppliers(sdeltaTime, svalue, size);
	}

}

File: src/main/scala/llmTests/a201701b/scala/gpt4/TraceFactoryImpl.scala
package llmTests.a201701b.scala.gpt4

import llmTests.a201701b.scala.Event
import llmTests.a201701b.scala.Trace
import llmTests.a201701b.scala.TraceFactory
import llmTests.a201701b.scala.gpt4.{EventImpl, TraceImpl}

import java.util.function.Supplier
import java.util.{ArrayList, Iterator, Optional}

class TraceFactoryImpl extends TraceFactory {
  override def fromSuppliers[X](sdeltaTime: Supplier[Integer], svalue: Supplier[X], size: Int): Trace[X] = {
    val events = new ArrayList[Event[X]](size)
    var time = 0
    for (_ <- 0 until size) {
      events.add(new EventImpl[X](time, svalue.get()))
      time += sdeltaTime.get()
    }
    new TraceImpl[X](events)
  }
  override def constant[X](sdeltaTime: Supplier[Integer], value: X, size: Int): Trace[X] = {
    val svalue = new Supplier[X]() { override def get(): X = value }
    fromSuppliers(sdeltaTime, svalue, size)
  }
  override def discrete[X](svalue: Supplier[X], size: Int): Trace[X] = {
    val sdeltaTime = new Supplier[Integer]() { override def get(): Integer = 1 }
    fromSuppliers(sdeltaTime, svalue, size)
  }
}

